pkg/awsiamauth/reconciler/reconciler.go

```diff
@@ -0,0 +1,123 @@
+type RemoteClientRegistry interface {
+	GetClient(ctx context.Context, cluster client.ObjectKey) (client.Client, error)
+}
+
+// Reconciler allows to reconcile AWSIamConfig.
+type Reconciler struct {
+	certgen              crypto.CertificateGenerator
+	templateBuilder      *awsiamauth.TemplateBuilder
+	clusterID            uuid.UUID
+	remoteClientRegistry RemoteClientRegistry
+}
+
+// New returns a new Reconciler.
+func New(certgen crypto.CertificateGenerator, clusterID uuid.UUID, remoteClientRegistry RemoteClientRegistry) *Reconciler {
+	return &Reconciler{
+		certgen:              certgen,
+		templateBuilder:      &awsiamauth.TemplateBuilder{},
+		clusterID:            clusterID,
+		remoteClientRegistry: remoteClientRegistry,
+	}
+}
+
+// ReconcileAWSIAMAuthCASecret ensures the AWS IAM Authenticator secret is present.
+// It uses a controller.Result to indicate when requeues are needed.
+func (r *Reconciler) ReconcileAWSIAMAuthCASecret(ctx context.Context, logger logr.Logger, client client.Client, clusterName string) (controller.Result, error) {
+	// Fetch the CA Secret
+	s := &corev1.Secret{}
+	err := client.Get(ctx, types.NamespacedName{Name: awsiamauth.GetAwsIamAuthCaSecretName(clusterName), Namespace: constants.EksaSystemNamespace}, s)
+	if apierrors.IsNotFound(err) {
+		logger.Info("Applying aws-iam-authenticator CA secret")
+		return r.applyCASecret(ctx, client, clusterName)
+	}
+	if err != nil {
+		return controller.Result{}, fmt.Errorf("fetching secret %s: %v", awsiamauth.GetAwsIamAuthCaSecretName(clusterName), err)
+	}
+
+	// If Secret is present, no-op
+	logger.Info("aws-iam-authenticator CA secret found", "Name", awsiamauth.GetAwsIamAuthCaSecretName(clusterName))
+	return controller.Result{}, nil
+}
+
+func (r *Reconciler) applyCASecret(ctx context.Context, client client.Client, clusterName string) (controller.Result, error) {
+	// Generate k8s secret yaml for aws-iam-authenticator
+	yaml, err := r.templateBuilder.GenerateCertKeyPairSecret(r.certgen, clusterName)
+	if err != nil {
+		return controller.Result{}, fmt.Errorf("generating aws-iam-authenticator ca secret: %v", err)
+	}
+
+	return controller.Result{}, serverside.ReconcileYaml(ctx, client, yaml)
+}
+
+// Reconcile takes the AWS IAM Authenticator indentity provider to the desired state defined in AWSIAMConfig.
+// It uses a controller.Result to indicate when requeues are needed.
+// Intended to be used in a kubernetes controller.
+func (r *Reconciler) Reconcile(ctx context.Context, logger logr.Logger, client client.Client, cluster *anywherev1.Cluster) (controller.Result, error) {
+	// Build the cluster spec for the Cluster
+	clusterSpec, err := anywhereCluster.BuildSpec(ctx, clientutil.NewKubeClient(client), cluster)
+	if err != nil {
+		return controller.Result{}, err
+	}
+
+	// Get the client object for the workload cluster
+	rClient, err := r.remoteClientRegistry.GetClient(ctx, controller.CapiClusterObjectKey(cluster))
```

Comment:
Before doing this, I would check if the CP is ready. This line will fail if not, so the end result would be similar, but it's a better way of handling such scenario.

Reply:
So currently the point in the code where this is called is after the workload cluster has come up. Since we requeue from the cluster controller till that.
But I can add a check to be sure 👍